File: caseforge/backend/src/services/codeRunner.js
const { execFile } = require("child_process");
const fs = require("fs");
const path = require("path");
const os = require("os");

function runPythonCode(code, timeout = 10) {
  return new Promise((resolve) => {
    const tmpFile = path.join(os.tmpdir(), `caseforge_${Date.now()}.py`);
    fs.writeFileSync(tmpFile, code);

    execFile(
      "python",
      [tmpFile],
      { timeout: timeout * 1000, env: { ...process.env, PYTHONDONTWRITEBYTECODE: "1" } },
      (error, stdout, stderr) => {
        // Clean up temp file
        try {
          fs.unlinkSync(tmpFile);
        } catch (_) {}

        if (error && error.killed) {
          resolve({
            stdout: "",
            stderr: `Execution timed out after ${timeout} seconds.`,
            exit_code: -1,
          });
        } else if (error) {
          resolve({
            stdout: stdout || "",
            stderr: stderr || error.message,
            exit_code: error.code || -1,
          });
        } else {
          resolve({
            stdout: stdout || "",
            stderr: stderr || "",
            exit_code: 0,
          });
        }
      }
    );
  });
}

module.exports = { runPythonCode };

File: caseforge/backend/services/code_runner.py
import subprocess
import tempfile
import os


def run_python_code(code: str, timeout: int = 10) -> dict:
    """Execute Python code in a subprocess with timeout.

    Returns dict with stdout, stderr, and exit_code.
    """
    with tempfile.NamedTemporaryFile(
        mode="w", suffix=".py", delete=False
    ) as f:
        f.write(code)
        f.flush()
        tmp_path = f.name

    try:
        result = subprocess.run(
            ["python", tmp_path],
            capture_output=True,
            text=True,
            timeout=timeout,
            env={**os.environ, "PYTHONDONTWRITEBYTECODE": "1"},
        )
        return {
            "stdout": result.stdout,
            "stderr": result.stderr,
            "exit_code": result.returncode,
        }
    except subprocess.TimeoutExpired:
        return {
            "stdout": "",
            "stderr": f"Execution timed out after {timeout} seconds.",
            "exit_code": -1,
        }
    except Exception as e:
        return {
            "stdout": "",
            "stderr": str(e),
            "exit_code": -1,
        }
    finally:
        try:
            os.unlink(tmp_path)
        except OSError:
            pass
